mist-core/src/main/java/edu/snu/mist/core/task/DefaultConfigDagGeneratorImpl.java

```diff
@@ -60,25 +58,50 @@ private DefaultConfigDagGeneratorImpl() {
     }
   }
 
-  /**
-   * Generate a dag that holds the configuration of vertices from avro vertex dag.
-   * @param avroDag avro vertex dag
-   * @return configuration vertex dag
-   */
   @Override
   public DAG<ConfigVertex, MISTEdge> generate(final AvroDag avroDag) {
+    return generateWithCheckpointedStates(avroDag, null);
+  }
+
+  @Override
+  public DAG<ConfigVertex, MISTEdge> generateWithCheckpointedStates(final AvroDag avroDag,
+                                                                    final QueryCheckpoint checkpointedState) {
     final List<ConfigVertex> deserializedVertices = new ArrayList<>(avroDag.getAvroVertices().size());
     final DAG<ConfigVertex, MISTEdge> configDag = new AdjacentListDAG<>();
+    final List<AvroVertex> avroVertices = avroDag.getAvroVertices();
 
-    // Fetch configurations from avro vertex dag
-    for (final AvroVertex avroVertex : avroDag.getAvroVertices()) {
-      final ExecutionVertex.Type type = getVertexType(avroVertex);
+    if (checkpointedState == null) {
+      // No checkpointed states discovered -> generate vertices without states
+      // Fetch configurations from avro vertex dag
+      for (final AvroVertex avroVertex : avroVertices) {
+        final ExecutionVertex.Type type = getVertexType(avroVertex);
 
-      // Create a config vertex
-      final ConfigVertex configVertex =
-          new ConfigVertex(Long.toString(configVertexId.getAndIncrement()), type, avroVertex.getConfiguration());
-      deserializedVertices.add(configVertex);
-      configDag.addVertex(configVertex);
+        // Create a config vertex
+        final ConfigVertex configVertex =
+            new ConfigVertex(Long.toString(configVertexId.getAndIncrement()), type, avroVertex.getConfiguration());
+        deserializedVertices.add(configVertex);
+        configDag.addVertex(configVertex);
+      }
+    } else {
+      // There are checkpointed states -> generate vertices with internal states.
+      final int numVertices = avroVertices.size();
+      final List<StateWithTimestamp> queryState = checkpointedState.getQueryState();
```

Comment:
Do we guarantee that the state order and vertex order are the same?

Reply:
No. Current `DAG` interface returns the set of vertices, which means that the order of the vertices can be changed according to its implementation. We need to give the separate id for each vertex to solve this problem.